File: src/xuezh/core/ids.py
from __future__ import annotations

import hashlib
import re

import ulid

# ID regexes (authoritative in specs/id-scheme.md)
WORD_ID_RE = re.compile(r"^w_[0-9a-f]{12}$")
GRAMMAR_ID_RE = re.compile(r"^g_[0-9a-f]{12}$")
CHAR_ID_RE = re.compile(r"^c_[0-9a-f]{12}$")
ITEM_ID_RE = re.compile(r"^[wgc]_[0-9a-f]{12}$")
CONTENT_ID_RE = re.compile(r"^ct_[0-9a-f]{12}$")
ARTIFACT_ID_RE = re.compile(r"^ar_[0-9a-f]{12}$")
EVENT_ULID_RE = re.compile(r"^ev_[0-9A-Z]{26}$")
EVENT_UUID_RE = re.compile(r"^ev_[0-9a-f]{32}$")


def normalize_pinyin(value: str) -> str:
    parts = value.strip().split()
    return " ".join(parts).lower()


def _hex12(payload: str) -> str:
    return hashlib.sha1(payload.encode("utf-8")).hexdigest()[:12]


def word_id(*, hanzi: str, pinyin: str) -> str:
    normalized = normalize_pinyin(pinyin)
    return f"w_{_hex12(f'word|{hanzi}|{normalized}')}"


def grammar_id(*, grammar_key: str) -> str:
    return f"g_{_hex12(f'grammar|{grammar_key}')}"


def char_id(*, character: str) -> str:
    return f"c_{_hex12(f'char|{character}')}"


def content_id(*, content_type: str, key: str) -> str:
    return f"ct_{_hex12(f'{content_type}|{key}')}"


def artifact_id(*, path: str) -> str:
    return f"ar_{_hex12(path)}"


def event_id_ulid() -> str:
    return f"ev_{ulid.new()}"


def is_word_id(value: str) -> bool:
    return bool(WORD_ID_RE.fullmatch(value))


def is_grammar_id(value: str) -> bool:
    return bool(GRAMMAR_ID_RE.fullmatch(value))


def is_char_id(value: str) -> bool:
    return bool(CHAR_ID_RE.fullmatch(value))


def is_item_id(value: str) -> bool:
    return bool(ITEM_ID_RE.fullmatch(value))


def is_event_id(value: str) -> bool:
    return bool(EVENT_ULID_RE.fullmatch(value) or EVENT_UUID_RE.fullmatch(value))


def is_content_id(value: str) -> bool:
    return bool(CONTENT_ID_RE.fullmatch(value))


def is_artifact_id(value: str) -> bool:
    return bool(ARTIFACT_ID_RE.fullmatch(value))

File: internal/xuezh/ids/ids.go
package ids

import (
	"crypto/rand"
	"crypto/sha1"
	"encoding/hex"
	"regexp"
	"strings"
	"time"

	"github.com/oklog/ulid/v2"
)

var (
	wordIDRe     = regexp.MustCompile(`^w_[0-9a-f]{12}$`)
	grammarIDRe  = regexp.MustCompile(`^g_[0-9a-f]{12}$`)
	charIDRe     = regexp.MustCompile(`^c_[0-9a-f]{12}$`)
	itemIDRe     = regexp.MustCompile(`^[wgc]_[0-9a-f]{12}$`)
	contentIDRe  = regexp.MustCompile(`^ct_[0-9a-f]{12}$`)
	artifactIDRe = regexp.MustCompile(`^ar_[0-9a-f]{12}$`)
	eventULIDRe  = regexp.MustCompile(`^ev_[0-9A-Z]{26}$`)
	eventUUIDRe  = regexp.MustCompile(`^ev_[0-9a-f]{32}$`)
)

var entropy = ulid.Monotonic(rand.Reader, 0)

func NormalizePinyin(value string) string {
	parts := strings.Fields(value)
	return strings.ToLower(strings.Join(parts, " "))
}

func hex12(payload string) string {
	h := sha1.Sum([]byte(payload))
	return hex.EncodeToString(h[:])[:12]
}

func WordID(hanzi, pinyin string) string {
	normalized := NormalizePinyin(pinyin)
	return "w_" + hex12("word|"+hanzi+"|"+normalized)
}

func GrammarID(grammarKey string) string {
	return "g_" + hex12("grammar|"+grammarKey)
}

func CharID(character string) string {
	return "c_" + hex12("char|"+character)
}

func ContentID(contentType, key string) string {
	return "ct_" + hex12(contentType+"|"+key)
}

func ArtifactID(path string) string {
	return "ar_" + hex12(path)
}

func EventIDULID() string {
	id := ulid.MustNew(ulid.Timestamp(time.Now()), entropy)
	return "ev_" + id.String()
}

func IsWordID(value string) bool {
	return wordIDRe.MatchString(value)
}

func IsGrammarID(value string) bool {
	return grammarIDRe.MatchString(value)
}

func IsCharID(value string) bool {
	return charIDRe.MatchString(value)
}

func IsItemID(value string) bool {
	return itemIDRe.MatchString(value)
}

func IsEventID(value string) bool {
	return eventULIDRe.MatchString(value) || eventUUIDRe.MatchString(value)
}

func IsContentID(value string) bool {
	return contentIDRe.MatchString(value)
}

func IsArtifactID(value string) bool {
	return artifactIDRe.MatchString(value)
}

func ItemType(value string) string {
	if IsWordID(value) {
		return "word"
	}
	if IsGrammarID(value) {
		return "grammar"
	}
	if IsCharID(value) {
		return "character"
	}
	return ""
}
